cognite/client/_api/sequences.py

```diff
@@ -692,15 +695,73 @@ def upsert(
+            cdf_item = self.retrieve(external_id=item.external_id)
+            if cdf_item and cdf_item.external_id:
+                cdf_item_by_id: dict[str, Sequence] = {cdf_item.external_id: cdf_item}
+        elif isinstance(item, collections.abc.Sequence):
+            external_ids = [i.external_id for i in item if isinstance(i, SequenceWrite)]
+            if None in external_ids:
+                raise ValueError("External ID must be set when upserting a SequenceWrite object.")
+            cdf_items = self.retrieve_multiple(external_ids=typing.cast(list[str], external_ids))
+            cdf_item_by_id = {cdf_item.external_id: cdf_item for cdf_item in cdf_items if cdf_item.external_id}
+        else:
+            cdf_item_by_id = {}
         return self._upsert_multiple(
             item,
             list_cls=SequenceList,
             resource_cls=Sequence,
             update_cls=SequenceUpdate,
             input_resource_cls=Sequence,
             mode=mode,
+            # String is not recognized as hashable by mypy
+            cdf_item_by_id=cdf_item_by_id,  # type: ignore[arg-type]
         )
 
+    @classmethod
+    def _convert_resource_to_patch_object(
+        cls,
+        resource: CogniteResource,
+        update_attributes: list[PropertySpec],
+        mode: Literal["replace_ignore_null", "patch", "replace"] = "replace_ignore_null",
+        cdf_item_by_id: dict[str, Sequence] | None = None,  # type: ignore[override]
+    ) -> dict[str, dict[str, dict]]:
+        update_obj = super()._convert_resource_to_patch_object(resource, update_attributes, mode)
+        if not isinstance(resource, SequenceWrite):
+            return update_obj
+        # Lookup columns to check what to add, remove and modify
+        cdf_item: Sequence | None = None
+        if cdf_item_by_id and resource.external_id:
+            cdf_item = cdf_item_by_id.get(resource.external_id)
+        update_obj["update"]["columns"] = {}
+        if cdf_item is None:
+            update_obj["update"]["columns"]["add"] = [column.dump() for column in resource.columns]
+        else:
+            cdf_columns_by_external_id = {
+                column.external_id: column for column in cdf_item.columns or [] if column.external_id
+            }
+            columns_by_external_id = {column.external_id: column for column in resource.columns if column.external_id}
+            if to_remove := (set(cdf_columns_by_external_id.keys()) - set(columns_by_external_id.keys())):
+                update_obj["update"]["columns"]["remove"] = list(to_remove)
+            if to_add := (set(columns_by_external_id.keys()) - set(cdf_columns_by_external_id.keys())):
+                update_obj["update"]["columns"]["add"] = [columns_by_external_id[ext_id].dump() for ext_id in to_add]
+            if to_modify := (set(columns_by_external_id.keys()) & set(cdf_columns_by_external_id.keys())):
```

Comment:
This doesn't seem to distinguish based on `mode` (`"replace_ignore_null", "patch", "replace"`)?

Reply:
You are right, good catch. Since columns are not nullable, I think `replace` and 'replace_ignore_null' should be the same behavior. While patch should not do the remove.